File: internal/canvaslms/graph.go
package canvaslms

import (
	"time"
)

const GraphQuery = `query gradeQuery {
  allCourses {
	_id
    id
    name
    state
	courseCode
    submissionsConnection(first: $maxn$, orderBy: {field: gradedAt, direction: descending}) {
      nodes {
        _id
        id
        assignment {
          _id
          id
          name
          dueAt
          gradingType
          pointsPossible
		  htmlUrl
        }
        score
        enteredScore
        grade
        enteredGrade
        gradingStatus
        gradeHidden
        gradedAt
        posted
		postedAt
		state
		user {
			_id
			id
			name
			sisId
			email
		}
      }
    }
  }
}`

type GraphResponse struct {
	Data struct {
		AllCourses []struct {
			IDLegacy              string `json:"_id"`
			ID                    string `json:"id"`
			Name                  string `json:"name"`
			State                 string `json:"state"`
			CourseCode            string `json:"courseCode"`
			SubmissionsConnection struct {
				Nodes []GraphSubmissionResponse `json:"nodes"`
			} `json:"submissionsConnection"`
		} `json:"allCourses"`
	} `json:"data"`
}
type GraphSubmissionResponse struct {
	IDLegacy   string `json:"_id"`
	ID         string `json:"id"`
	Assignment struct {
		IDLegacy       string  `json:"_id"`
		ID             string  `json:"id"`
		Name           string  `json:"name"`
		DueAt          *string `json:"dueAt"`
		GradingType    string  `json:"gradingType"`
		PointsPossible float64 `json:"pointsPossible"`
		HTMLUrl        string  `json:"htmlUrl"`
	} `json:"assignment"`
	Score         *float64 `json:"score"`
	EnteredScore  *float64 `json:"enteredScore"`
	Grade         *string  `json:"grade"`
	EnteredGrade  *string  `json:"enteredGrade"`
	GradingStatus string   `json:"gradingStatus"`
	GradeHidden   bool     `json:"gradeHidden"`
	GradedAt      *string  `json:"gradedAt"`
	Posted        bool     `json:"posted"`
	PostedAt      *string  `json:"postedAt"`
	State         string   `json:"state"`
	User          struct {
		IDLegacy string  `json:"_id"`
		ID       string  `json:"id"`
		SISID    *string `json:"sisId"`
		Name     string  `json:"name"`
		Email    *string `json:"email"`
	}
}

type GraphSubmissionCompareFunc func(m1, m2 GraphSubmissionResponse) (m1HasPriority bool)

func parseJSONTime(s string) time.Time {
	t, _ := time.Parse(time.RFC3339, s)
	return t
}

func GraphSubmissionCompareByDue(m1, m2 GraphSubmissionResponse) (m1HasPriority bool) {
	if m1.Assignment.DueAt == nil {
		return false
	}
	if m2.Assignment.DueAt == nil {
		return true
	}

	m1Time, m2Time := parseJSONTime(*m1.Assignment.DueAt), parseJSONTime(*m2.Assignment.DueAt)
	now := time.Now()
	m1IsPast, m2IsPast := now.After(m1Time), now.After(m2Time)
	if m1IsPast && m2IsPast {
		return m1Time.After(m2Time)
	}
	if !m1IsPast && !m2IsPast {
		return m1Time.Before(m2Time)
	}
	return !m1IsPast
}

func laterTime(t1, t2 *string) *string {
	if t1 == nil {
		return t2
	}
	if t2 == nil {
		return t1
	}
	t1T, t2T := parseJSONTime(*t1), parseJSONTime(*t2)

	if t1T.After(t2T) {
		return t1
	}
	return t2
}

func GraphSubmissionCompareByGradeTime(m1, m2 GraphSubmissionResponse) (m1HasPriority bool) {
	m1LastUpdate := laterTime(m1.PostedAt, m1.GradedAt)
	m2LastUpdate := laterTime(m2.PostedAt, m2.GradedAt)
	if m2LastUpdate == nil {
		return true
	}
	if m1LastUpdate == nil {
		return false
	}
	return parseJSONTime(*m1LastUpdate).After(parseJSONTime(*m2LastUpdate))
}

File: src/apps/canvas_lms/graph.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize)]
pub struct GraphQuery<T> {
    pub query: String,
    #[serde(rename = "operationName")]
    pub operation_name: String,
    pub variables: T,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct GraphNodes<T> {
    pub nodes: Vec<T>,
}

#[derive(Debug, Clone, Deserialize)]
pub struct GraphResponse<T> {
    pub data: T,
}

pub const ALL_COURSES_QUERY: &str = "query gradeQuery {
    allCourses {
      _id
      id
      name
      state
      courseCode
      submissionsConnection(first: $maxn$, orderBy: {field: gradedAt, direction: descending}) {
        nodes {
          _id
          id
          assignment {
            _id
            id
            name
            dueAt
            gradingType
            pointsPossible
            htmlUrl
          }
          score
          enteredScore
          grade
          enteredGrade
          gradingStatus
          gradeHidden
          gradedAt
          posted
          postedAt
          state
          user {
              _id
              id
              name
              sisId
              email
          }
        }
      }
    }
  }";

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct AllCourseData {
    #[serde(rename = "allCourses")]
    pub all_courses: Vec<GraphCourse>,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct GraphCourse {
    #[serde(rename = "_id")]
    pub id_legacy: String,
    pub id: String,
    pub name: String,
    pub state: String,
    #[serde(rename = "courseCode")]
    pub course_code: String,
    #[serde(rename = "submissionsConnection")]
    pub submissions_connection: GraphNodes<GraphSubmission>,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct GraphSubmission {
    #[serde(rename = "_id")]
    pub id_legacy: String,
    pub id: String,
    pub assignment: GraphAssignment,
    pub score: Option<f64>,
    #[serde(rename = "enteredScore")]
    pub entered_score: Option<f64>,
    pub grade: Option<String>,
    #[serde(rename = "enteredGrade")]
    pub entered_grade: Option<String>,
    #[serde(rename = "gradingStatus")]
    pub grading_status: Option<String>,
    #[serde(rename = "gradeHidden")]
    pub grade_hidden: bool,
    #[serde(rename = "gradedAt")]
    pub graded_at: Option<chrono::DateTime<chrono::FixedOffset>>,
    pub posted: bool,
    #[serde(rename = "postedAt")]
    pub posted_at: Option<chrono::DateTime<chrono::FixedOffset>>,
    pub state: String,
    pub user: GraphUser,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct GraphAssignment {
    #[serde(rename = "_id")]
    pub id_legacy: String,
    pub id: String,
    pub name: String,
    #[serde(rename = "dueAt")]
    pub due_at: Option<chrono::DateTime<chrono::FixedOffset>>,
    #[serde(rename = "gradingType")]
    pub grading_type: String,
    #[serde(rename = "pointsPossible")]
    pub points_possible: f64,
    #[serde(rename = "htmlUrl")]
    pub html_url: String,
}

#[derive(Debug, Clone, Deserialize, Serialize)]
pub struct GraphUser {
    #[serde(rename = "_id")]
    pub id_legacy: String,
    pub id: String,
    #[serde(rename = "sisId")]
    pub sis_id: Option<String>,
    pub name: String,
    pub email: Option<String>,
}
